### PR TITLE
[DLv2] Make graph `traverse` working with unhashable `DataPipe`

### PR DESCRIPTION
**Pull request by ejguan**
Summary:
This Diff removes the requirement for `traverse` function that `DataPipe` needs to be hash-able. `traverse` function now is using `id` of `DataPipe` instance rather than `DataPipe` itself as the key for both `cache` and graph.

But, it requires the changes of type of `DataPipeGraph` from `Dict[DataPipe, "DataPipeGraph"]` to `Dict[int, Tuple[DataPipe, "DataPipeGraph"]]`.

Differential Revision: D37354153

Ref PR in TorchData: https://github.com/pytorch/data/pull/559

### CODE REVIEW

File: torch/utils/data/graph.py [hunk 1]
```diff
@@ -4,11 +4,13 @@
 from torch.utils.data import IterDataPipe, MapDataPipe
 from torch.utils.data._utils.serialization import DILL_AVAILABLE
 
-from typing import Any, Dict, Set, Tuple, Type, Union
+from typing import Dict, List, Set, Tuple, Type, Union
 
-__all__ = ["traverse", ]
+__all__ = ["DataPipe", "DataPipeGraph", "traverse"]
 
 DataPipe = Union[IterDataPipe, MapDataPipe]
```
Comment (ejguan) [inline torch/utils/data/graph.py:11]: Should I rename it to `DataPipeType`?

File: torch/utils/data/graph.py [hunk 2]
```diff
@@ -70,21 +72,22 @@ def reduce_hook(obj):
...
 
 # Add cache here to prevent infinite recursion on DataPipe
-def _traverse_helper(datapipe, only_datapipe, cache):
+def _traverse_helper(datapipe: DataPipe, only_datapipe: bool, cache: Set[int]) -> DataPipeGraph:
     if not isinstance(datapipe, (IterDataPipe, MapDataPipe)):
         raise RuntimeError("Expected `IterDataPipe` or `MapDataPipe`, but {} is found".format(type(datapipe)))
 
-    cache.add(datapipe)
+    dp_id = id(datapipe)
+    cache.add(id(dp_id))
```
Comment (NivekT) [inline torch/utils/data/graph.py:86]:
```suggestion
    cache.add(dp_id)
```
I think this is what you want and should fix the infinite recursion in CI.

File: torch/utils/data/graph.py (hunk 1 above)
Comment (NivekT) [reply torch/utils/data/graph.py:11]: I am indifferent.

File: torch/utils/data/graph.py [hunk 3]
```diff
@@ -4,11 +4,13 @@
...
 from torch.utils.data._utils.serialization import DILL_AVAILABLE
 
-from typing import Any, Dict, Set, Tuple, Type, Union
+from typing import Dict, List, Set, Tuple, Type, Union
 
-__all__ = ["traverse", ]
+__all__ = ["DataPipe", "DataPipeGraph", "traverse"]
 
 DataPipe = Union[IterDataPipe, MapDataPipe]
+DataPipeGraph = Dict[int, Tuple[DataPipe, "DataPipeGraph"]]  # type: ignore[misc]
```
Comment (NivekT) [inline torch/utils/data/graph.py:12]: Do their `__module__` need to change or they need to be added to `__init__` to fix the following CI error?
https://github.com/pytorch/pytorch/runs/7114135518?check_suite_focus=true

**Review by NivekT — COMMENTED** [review]
Mostly LGTM! Needs to fix the CI issues

**Review by VitalyFedyunin — APPROVED** [review]

**Pull request closed without merging**